go/vt/vtgate/planbuilder/operators/delete.go

```diff
@@ -144,18 +132,78 @@ func deleteWithInputPlanningForFk(ctx *plancontext.PlanningContext, del *sqlpars
 		Limit:   delClone.Limit,
 		Lock:    sqlparser.ForUpdateLock,
 	}
-	ts := ctx.SemTable.Targets[del.Targets[0].Name]
+
+	var delOps []delOp
+	for _, target := range del.Targets {
+		op := createDeleteOpWithTarget(ctx, target)
+		delOps = append(delOps, op)
+	}
+
+	// sort the operator based on sharding vindex type.
+	// Unsharded < Lookup Vindex < Any
+	sort.Slice(delOps, func(i, j int) bool {
+		a, b := delOps[i], delOps[j]
+		// Get the first Vindex of a and b, if available
+		aVdx, bVdx := getFirstVindex(a.vTbl), getFirstVindex(b.vTbl)
+
+		// Sort nil Vindexes to the start
+		if aVdx == nil || bVdx == nil {
+			return aVdx != nil // true if bVdx is nil and aVdx is not nil
+		}
+
+		// Among non-nil Vindexes, those that need VCursor come first
+		return aVdx.NeedsVCursor() && !bVdx.NeedsVCursor()
+	})
```

Comment:
Why is this required?

Reply:
We need to delete the tables based on the sharding vindex type.
Example
```
user - id is a sharding key, phone as lookup key)
directory - phone as lookup sharding key)
```

`delete u, d from user u join directory d on u.phone = d.phone`

Here if we delete rows from `user` table first then data will be delete from `phone` lookup table as well.
When we try to delete from `directory` table we will not be able to find the data in `phone` lookup table as it was already delete when deleting from `user` table. The delete query for `directory` will not be sent to any shard.

Reply:
Okay, I understand, might be worth adding a comment explaining this.